politeiawww/dcc.go

```diff
@@ -0,0 +1,608 @@
+	if err != nil {
+		return nil, err
+	}
+
+	// Setup politeiad request
+	challenge, err := util.Random(pd.ChallengeSize)
+	if err != nil {
+		return nil, err
+	}
+	n := pd.NewRecord{
+		Challenge: hex.EncodeToString(challenge),
+		Metadata: []pd.MetadataStream{
+			{
+				ID:      mdStreamDCCGeneral,
+				Payload: string(md),
+			},
+			{
+				ID:      mdStreamDCCStatusChanges,
+				Payload: string(scb),
+			},
+		},
+		Files: convertPropFilesFromWWW(nd.Files),
+	}
+
+	// Send the newrecord politeiad request
+	responseBody, err := p.makeRequest(http.MethodPost,
+		pd.NewRecordRoute, n)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Infof("Submitted issuance nomination: %v", u.Username)
+	for k, f := range n.Files {
+		log.Infof("%02v: %v %v", k, f.Name, f.Digest)
+	}
+
+	// Handle newRecord response
+	var pdReply pd.NewRecordReply
+	err = json.Unmarshal(responseBody, &pdReply)
+	if err != nil {
+		return nil, fmt.Errorf("Unmarshal NewDCCReply: %v", err)
+	}
+
+	// Verify NewRecord challenge
+	err = util.VerifyChallenge(p.cfg.Identity, challenge, pdReply.Response)
+	if err != nil {
+		return nil, err
+	}
+
+	// Change politeiad record status to public. DCCs
+	// do not need to be reviewed before becoming public.
+	// An admin signature is not included for this reason.
+	c := MDStreamChanges{
+		Version:   VersionMDStreamChanges,
+		Timestamp: time.Now().Unix(),
+		NewStatus: pd.RecordStatusPublic,
+	}
+	blob, err := encodeMDStreamChanges(c)
+	if err != nil {
+		return nil, err
+	}
+
+	challenge, err = util.Random(pd.ChallengeSize)
+	if err != nil {
+		return nil, err
+	}
+
+	sus := pd.SetUnvettedStatus{
+		Token:     pdReply.CensorshipRecord.Token,
+		Status:    pd.RecordStatusPublic,
+		Challenge: hex.EncodeToString(challenge),
+		MDAppend: []pd.MetadataStream{
+			{
+				ID:      mdStreamChanges,
+				Payload: string(blob),
+			},
+		},
+	}
+
+	// Send SetUnvettedStatus request to politeiad
+	responseBody, err = p.makeRequest(http.MethodPost,
+		pd.SetUnvettedStatusRoute, sus)
+	if err != nil {
+		return nil, err
+	}
+
+	var pdSetUnvettedStatusReply pd.SetUnvettedStatusReply
+	err = json.Unmarshal(responseBody, &pdSetUnvettedStatusReply)
+	if err != nil {
+		return nil, fmt.Errorf("Could not unmarshal SetUnvettedStatusReply: %v",
+			err)
+	}
+
+	// Verify the SetUnvettedStatus challenge.
+	err = util.VerifyChallenge(p.cfg.Identity, challenge,
+		pdSetUnvettedStatusReply.Response)
+	if err != nil {
+		return nil, err
+	}
+
+	r := pd.Record{
+		Metadata:         n.Metadata,
+		Files:            n.Files,
+		CensorshipRecord: pdReply.CensorshipRecord,
+	}
+	// Submit issuance to cmsdb
+
+	dccRec, err := convertRecordToDatabaseDCC(r)
+	if err != nil {
+		return nil, err
+	}
+
+	err = p.cmsDB.NewDCC(dccRec)
+	if err != nil {
+		return nil, err
+	}
+
+	cr := convertPropCensorFromPD(pdReply.CensorshipRecord)
+
+	reply.CensorshipRecord = cr
+	return reply, nil
+}
+
+func (p *politeiawww) validateDCC(nd cms.NewDCC, u *user.User) error {
```

Comment:
This function copies the proposal/invoice validation, but it should really be limited to just what is needed for a dcc.